shapely/strtree.py

```diff
@@ -66,78 +68,94 @@ class STRtree:
     >>> polys[2] in result
     False
 
-    Behavior if an `STRtree` is created empty:
+    Notes
+    -----
+    The class maintains a reverse mapping of integer items to geometries
+    and for performance reasons stores only integers in the STRtree from
+    the GEOS C API. The tree is filled using the Sort-Tile-Recursive
+    [1]_ algorithm.
+
+    References
+    ----------
+    .. [1] Leutenegger, Scott T.; Edgington, Jeffrey M.; Lopez, Mario A.
+    (February 1997). "STR: A Simple and Efficient Algorithm for R-Tree
+    Packing". https://ia600900.us.archive.org/27/items/nasa_techdoc_19970016975/19970016975.pdf
 
-    >>> tree = STRtree([])
-    >>> tree.query(Point(0, 0))
-    []
-    >>> print(tree.nearest(Point(0, 0)))
-    None
     """
 
-    def __init__(self, geoms):
+    def __init__(self, initdata):
         warn(
-            "STRtree will be completely changed in 2.0.0. The exact API is not yet decided, but will be documented before 1.8.0",
+            "STRtree will be changed in 2.0.0. The exact API is not yet decided, but will be documented before 1.8.0",
             ShapelyDeprecationWarning,
             stacklevel=2,
         )
-        # filter empty geometries out of the input
-        geoms = [geom for geom in geoms if not geom.is_empty]
-        self._n_geoms = len(geoms)
-
-        self._init_tree_handle(geoms)
-
-        # Keep references to geoms.
-        self._geoms = list(geoms)
-
-    def _init_tree_handle(self, geoms):
-        node_capacity = 10
-        self._tree_handle = lgeos.GEOSSTRtree_create(node_capacity)
-        for geom in geoms:
-            lgeos.GEOSSTRtree_insert(self._tree_handle, geom._geom, ctypes.py_object(geom))
+        self._tree = None
+        self._rev = None
+        if initdata:
+            self._rev = {
+                idx: geom
+                for geom, idx in self._iterinitdata(initdata)
+                if not geom.is_empty
+            }
+            if self._rev:
+                self._init_tree(self._rev.items())
+
+    def _iterinitdata(self, initdata):
+        for enum_idx, item in enumerate(initdata):
+            if isinstance(item, tuple):  # a geom, idx pair
+                yield item[:2]
+            else:  # a single geometry
+                yield (item, enum_idx)
```

Comment:
Here I think we would need to add some validation that the user doesn't mix both cases (otherwise a user-provided id could clash with a `enum_idx` of another geometry)

(it would of course be very bizarre for a user to do so, but I think still worth validating in some way)

Reply:
@jorisvandenbossche this is a "private" method called only from the constructor, so when our constructor's code is correct we'll never see a mixed mode.

Reply:
Now it indeed is always called consistently from the constructor, but that was not the case when I commented :-) (then it was passing through the list of tuples from the user)